pkg/kubepkg/kubepkg.go

```diff
@@ -191,169 +241,135 @@ func buildPackage(pkg, arch string, packageDef *PackageDefinition) error {
 	if err != nil {
 		return errors.Wrap(err, "getting CNI download link")
 	}
 
-	logrus.Infof("Building %s package for %s/%s architecture...", c.Package, c.Arch, c.DebArch)
-	return c.run()
+	logrus.Infof("Building %s package for %s/%s architecture...", bc.Package, bc.GoArch, bc.BuildArch)
+	return bc.run()
 }
 
-func (c cfg) run() error {
-	var w []work
-
-	// nolint
-	// TODO: Get package directory for any version once package definitions are broken out
-	srcdir := filepath.Join(latestPackagesDir, c.Package)
-	dstdir, err := ioutil.TempDir(os.TempDir(), "debs")
+func (bc *buildConfig) run() error {
+	workspaceInfo, err := os.Stat(bc.workspace)
 	if err != nil {
 		return err
 	}
 
-	if !*keepTmp {
-		defer os.RemoveAll(dstdir)
-	}
-
-	if err := filepath.Walk(srcdir, func(srcfile string, f os.FileInfo, err error) error {
-		if err != nil {
-			return err
-		}
-		dstfile := filepath.Join(dstdir, srcfile[len(srcdir):])
-		if dstfile == dstdir {
-			return nil
-		}
-		if f.IsDir() {
-			return os.Mkdir(dstfile, f.Mode())
-		}
-		t, err := template.
-			New("").
-			Funcs(builtins).
-			Option("missingkey=error").
-			ParseFiles(srcfile)
-		if err != nil {
-			return err
-		}
-		w = append(w, work{
-			src:  srcfile,
-			dst:  dstfile,
-			t:    t.Templates()[0],
-			info: f,
-		})
+	specDir := filepath.Join(bc.workspace, string(bc.Channel), bc.Package)
+	specDirWithArch := filepath.Join(specDir, bc.GoArch)
 
-		return nil
-	}); err != nil {
+	err = os.MkdirAll(specDirWithArch, workspaceInfo.Mode())
+	if err != nil {
 		return err
 	}
 
-	for _, w := range w {
-		if err := func() error {
-			f, err := os.OpenFile(w.dst, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0)
-			if err != nil {
-				return err
-			}
-			defer f.Close()
+	//nolint:godox
+	// TODO: keepTmp/cleanup needs to defined in kubepkg root
+	if !bc.specOnly {
+		defer os.RemoveAll(specDirWithArch)
+	}
 
-			if err := w.t.Execute(f, c); err != nil {
-				return err
-			}
-			if err := os.Chmod(w.dst, w.info.Mode()); err != nil {
-				return err
-			}
-			return nil
-		}(); err != nil {
-			return err
-		}
+	_, err = buildSpecs(bc, specDirWithArch)
+	if err != nil {
+		return err
 	}
 
-	logrus.Infof("Running dpkg-buildpackage for %s (%s/%s)", c.Package, c.Arch, c.DebArch)
+	if bc.specOnly {
+		logrus.Info("spec-only mode was selected")
+		return nil
+	}
 
-	dpkgStatus, dpkgErr := command.NewWithWorkDir(
-		dstdir,
-		"dpkg-buildpackage",
-		"--unsigned-source",
-		"--unsigned-changes",
-		"--build=binary",
-		"--host-arch",
-		c.DebArch,
-	).Run()
+	//nolint:godox
+	// TODO: Move OS-specific logic into their own files
+	switch bc.Type {
+	case BuildDeb:
+		logrus.Infof("Running dpkg-buildpackage for %s (%s/%s)", bc.Package, bc.GoArch, bc.BuildArch)
+
+		dpkgStatus, dpkgErr := command.NewWithWorkDir(
+			specDirWithArch,
+			"dpkg-buildpackage",
+			"--unsigned-source",
+			"--unsigned-changes",
+			"--build=binary",
+			"--host-arch",
+			bc.BuildArch,
+		).Run()
```

Comment:
We can use `RunSuccess()` here which automatically checks for the right return type.